File: go-terrapod/encryption.go
package terrapod

import (
	"context"
	"encoding/json"
)

// EncryptionStatus reports application-layer encryption-at-rest health (#553).
// Decryptable is the headline durability signal — false means the platform
// cannot currently read its encrypted data back (investigate immediately).
type EncryptionStatus struct {
	Enabled       bool   `json:"enabled"`
	Provider      string `json:"provider"`
	ActiveVersion *int   `json:"active_version"`
	DEKVersions   []int  `json:"dek_versions"`
	CanaryOK      bool   `json:"canary_ok"`
	Decryptable   bool   `json:"decryptable"`
}

// GetEncryptionStatus fetches encryption-at-rest status. Requires platform admin.
func (c *Client) GetEncryptionStatus(ctx context.Context) (*EncryptionStatus, error) {
	body, err := c.Get(ctx, "/api/terrapod/v1/admin/encryption")
	if err != nil {
		return nil, err
	}
	var env struct {
		Data struct {
			Attributes EncryptionStatus `json:"attributes"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &env); err != nil {
		return nil, err
	}
	return &env.Data.Attributes, nil
}

File: services/terrapod/api/routers/encryption.py
"""Encryption-at-rest status endpoint (admin only).

UX/consumer contract: consumed by go-terrapod (GetEncryptionStatus) and surfaced
read-only. Reports whether app-layer encryption is enabled and — the headline
durability signal — whether everything is currently **decryptable**. Read-only;
no secrets or key material are ever returned.
"""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from terrapod.api.dependencies import AuthenticatedUser, require_admin
from terrapod.crypto.service import get_encryption, rotate_dek
from terrapod.db.session import get_db

router = APIRouter(prefix="/admin", tags=["encryption"])


@router.get("/encryption")
async def get_encryption_status(
    user: AuthenticatedUser = Depends(require_admin),
) -> dict:
    """Return encryption-at-rest health (admin only).

    ``decryptable`` is the key field: false means the platform booted but the
    active key can't read the canary back — investigate immediately
    (docs/encryption-at-rest.md#recovery).
    """
    return {"data": {"type": "encryption-status", "attributes": get_encryption().status()}}


@router.post("/encryption/rotate-dek")
async def rotate_data_encryption_key(
    user: AuthenticatedUser = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """Mint a new active DEK (admin only). Prior versions are retained so all
    existing ciphertext stays decryptable; run the encrypt migration to re-encrypt
    old rows under the new key. 409 when encryption is disabled.
    """
    try:
        version = await rotate_dek(db)
    except RuntimeError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    return {
        "data": {
            "type": "encryption-status",
            "attributes": {**get_encryption().status(), "rotated_to_version": version},
        }
    }
